### PR TITLE
GH-100923: Embed jump mask in `COMPARE_OP` oparg

### PR DESCRIPTION
**Pull request by markshannon**
Saves a few bytes in the code array.

* Issue: gh-100923

### CODE REVIEW

File: Lib/importlib/_bootstrap_external.py [hunk 1]
```diff
@@ -440,7 +441,7 @@ def _write_atomic(path, data, mode=0o666):
 # Whenever MAGIC_NUMBER is changed, the ranges in the magic_values array
 # in PC/launcher.c must also be updated.
 
-MAGIC_NUMBER = (3514).to_bytes(2, 'little') + b'\r\n'
+MAGIC_NUMBER = (3512).to_bytes(2, 'little') + b'\r\n'
```
Comment (iritkatriel) [inline Lib/importlib/_bootstrap_external.py]:
```suggestion
MAGIC_NUMBER = (3515).to_bytes(2, 'little') + b'\r\n'
```

File: Python/bytecodes.c [hunk 2]
```diff
@@ -1890,29 +1890,30 @@ dummy_func(
...
             STAT_INC(COMPARE_OP, hit);
             int res = _PyUnicode_Equal(left, right);
-            assert(oparg == Py_EQ || oparg == Py_NE);
+            assert((oparg >>4) == Py_EQ || (oparg >>4) == Py_NE);
             _Py_DECREF_SPECIALIZED(left, _PyUnicode_ExactDealloc);
             _Py_DECREF_SPECIALIZED(right, _PyUnicode_ExactDealloc);
             assert(res == 0 || res == 1);
-            assert(invert == 0 || invert == 1);
-            jump = res ^ invert;
+            assert((oparg & 15) == 7 || (oparg & 15) == 8);
```
Comment (iritkatriel) [inline Python/bytecodes.c]:
```suggestion
            assert((oparg & 0xF) == 7 || (oparg & 0xF) == 8);
```

Comment (iritkatriel) [reply Python/bytecodes.c]: What are 7 and 8?

Comment (markshannon) [reply Python/bytecodes.c]: Good question.
I've factored out the ordering to bit mask logic and given things names.

File: Python/compile.c [hunk 3]
```diff
@@ -2887,7 +2887,8 @@ static int compiler_addcompare(struct compiler *c, location loc,
     default:
         Py_UNREACHABLE();
     }
-    ADDOP_I(c, loc, COMPARE_OP, cmp);
+    /* cmp goes in top bits of the oparg, low bits are for the mask. */
```
Comment (iritkatriel) [inline Python/compile.c]:
```suggestion
    /* cmp goes in top bits of the oparg, while the low bits are used by specialized
     * versions of this opcode to store the comparison mask. */
```

**Review by iritkatriel — APPROVED** [review]

**Comment by sweeneyde** [comment]
Can the mask be added at compile time to avoid mutating the logical code?

```python
def f():
    ...
c0 = f.__code__.replace()
# ... run the code here ...
c1 = f.__code__

assert c0 == c1              # fails?
assert hash(c0) == hash(c1)  # fails?
```

**Comment by sweeneyde** [comment]
Oh I see, the work is being done in PyCode_Quicken, which is already called on every code object, so we should be okay.

**Comment by markshannon** [comment]
Yes, the plan is to move `PyCode_Quicken` into the compiler. https://github.com/faster-cpython/ideas/issues/530

File: Python/compile.c [hunk 4]
```diff
@@ -2887,7 +2887,9 @@ static int compiler_addcompare(struct compiler *c, location loc,
     default:
         Py_UNREACHABLE();
     }
-    ADDOP_I(c, loc, COMPARE_OP, cmp);
+    /* cmp goes in top bits of the oparg, while the low bits are used by specialized
```
Comment (iritkatriel) [inline Python/compile.c]:
```suggestion
    /* cmp goes in top bits of the oparg, while the low bits are used by quickened
```

**Comment by esc** [comment]
I just stumbled upon this. It seems like this PR didn't update the docs, as:

https://docs.python.org/3.12/library/dis.html#opcode-COMPARE_OP

Reads:

Screenshot 2023-09-25 at 18 24 05

Perhaps the new masking mechanism could be included here? Or maybe it will suffice to update the docs to say `cmp_op[opname>>4]`? Would be happy to try to submit a PR if that is welcome.

**Comment by brandtbucher** [comment]
That issue is being covered in #107457.

**Pull request merged**